apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
-/* global Promise,
-      EventDispatcher
-*/
-
-/* exported Navigation */
+/*global
+  Utils
+ */
+(function(exports) {
 
 'use strict';
 
-(function(window) {
+/**
+ * The Defer type is useful when creating promises.
+ * @typedef {Object} Defer
+ * @property {function(*)} resolve The Promise's resolve function.
+ * @property {function(*)} reject The Promise's reject function.
+ * @property {Promise} promise The actual promise.
+ */
+
+/**
+ * @function
+ * The debug function translates to console.log in debug mode.
+ *
+ * @param {...*} args The values to be displayed to the console.
+ */
+const debug = 1 ? (...args) => console.log(...args) : () => {};
+
+/**
+  * This contains an array of view characteristics, necessary to handle
+  * navigation requests.
+  * Each item has the following properties:
+  * - name: this is the name of the view. This will be used also in the URL
+  *   hash if the same HTML file holds several views.
+  * - url: this is the URL used to match the current location and find the
+  *   current view.
+  * - view: this is the actual JavaScript object that manages this view.
+  * - previous (optional): this is the name of the previous view; this is
```

Comment:
This is used only when doing in-document view slides, so not when doing between-document navigation.

Maybe we could record the panel changes and use that instead, and use "previous" only when we don't know the previous panel (for example because the user accessed a view directly and presses the "back" button that uses toPanel ?).

Or maybe "previous" could be an array ? but some cases would not work (like new_message <---> conversation). Need to think a little more :)

> conversation ---> new_message (doesn't exist now);

I think it does exist if there is a phone number in a message, and we use it to send a new message.